### PR TITLE
Feat: xml diff

### PR DESCRIPTION
**Pull request by dleard**
issue #268
- [ ] tests

### CODE REVIEW

**Comment by dleard** [comment]
Updated diff with styled summarize/swap buttons & syntax highlighting by tag depth
updatedDiff

File: ggircs-app/app/components/Dashboard/SwrsDataAccessCard.tsx [hunk 1]
```diff
@@ -11,8 +11,16 @@ const SwrsDataAccessCard: React.FunctionComponent = () => (
     <Link href="/swrs-browser" passHref>
       <Button marginTop="2em">Browse SWRS files</Button>
     </Link>
+    <div id="bottom-button">
+      <Link href="/xml-diff" passHref>
+        <Button marginTop="2em">Diff XML data</Button>
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/Dashboard/SwrsDataAccessCard.tsx:16]:
```suggestion
        <Button marginTop="2em">Compare SWRS report XML files</Button>
```

File: ggircs-app/app/components/XmlDiff/DiffDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import React from "react";
+import { Row, Col } from "react-bootstrap";
+
+interface Props {
+  report: any;
+}
+
+export const DiffDetails: React.FunctionComponent<Props> = ({ report }) => (
+  <>
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/XmlDiff/DiffDetails.tsx:9]: don't need this fragment

File: ggircs-app/app/components/XmlDiff/DiffDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from "react";
+import { Row, Col } from "react-bootstrap";
+
+interface Props {
+  report: any;
+}
+
+export const DiffDetails: React.FunctionComponent<Props> = ({ report }) => (
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/XmlDiff/DiffDetails.tsx:8]: Maybe we could have three `string` props instead, to avoid a complex object with `any` type

File: ggircs-app/app/components/XmlDiff/RenderDiff.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+          break;
+        case 20:
+          tagColor = "#8d6708";
+          break;
+        case 24:
+          tagColor = "#1978d4";
+          break;
+        default:
+          tagColor = "#ff0000";
+      }
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/XmlDiff/RenderDiff.tsx:83]: this could be written as an array, and you access the tag color by using `leadingWhiteSpace / 4` as the index

File: ggircs-app/app/components/XmlDiff/RenderDiff.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+export const RenderDiff: React.FunctionComponent<Props> = ({
+  oldText,
+  newText,
+  collapse,
+}) => {
+  const [loaded, setLoaded] = useState(false);
+
+  useEffect(() => {
+    setLoaded(true);
+  }, []);
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/XmlDiff/RenderDiff.tsx:24]: Why is this needed? Setting it immediately to `true` is a bit confusing. Is there a callback somewhere where `setLoaded(true)` would be more appropriate?

File: ggircs-app/app/pages/xml-diff.tsx [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  `;
+
+  state = {
+    leftSideId: null,
+    leftSideReport: null,
+    rightSideId: null,
+    rightSideReport: null,
+    isReversed: false,
+    isCollapsed: false,
+  };
```
Comment (matthieu-foucault) [inline ggircs-app/app/pages/xml-diff.tsx:75]: do we want to move those to the query string?

File: ggircs-app/app/next-env.d.ts [hunk 7]
```diff
@@ -22,9 +22,9 @@ export type PageComponent = NextComponentType<
   PageComponentProps
 > &
   ComponentClass<PageComponentProps> & {
-    static query: GraphQLTaggedNode;
-    static isAccessProtected: boolean;
-    static allowedGroups: string[];
+    query: GraphQLTaggedNode;
+    isAccessProtected: boolean;
+    allowedGroups: string[];
```
Comment (matthieu-foucault) [inline ggircs-app/app/next-env.d.ts:27]: Why this change?

File: ggircs-app/app/components/XmlDiff/ReportSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
+import React, { useState } from "react";
+import Input from "@button-inc/bcgov-theme/Input";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faCheck, faTimes } from "@fortawesome/free-solid-svg-icons";
+
+interface Props {
+  diffSide: String;
+  setSwrsReportId: (id: number, xml: any) => void;
+  swrsReportId?: number;
+  allReports: any;
```
Comment (matthieu-foucault) [inline ggircs-app/app/components/XmlDiff/ReportSelector.tsx:10]: A couple of `any` types here. Can we type them?

**Review by matthieu-foucault — CHANGES_REQUESTED** [review]

File: ggircs-app/app/next-env.d.ts (hunk 7 above)
Comment (dleard) [reply ggircs-app/app/next-env.d.ts:27]: it looks like pre-commit did this? I changed it back to include the static keywords and pre-commit auto-changed it again :upside_down_face:

File: ggircs-app/app/components/XmlDiff/RenderDiff.tsx (hunk 5 above)
Comment (dleard) [reply ggircs-app/app/components/XmlDiff/RenderDiff.tsx:24]: This is supposed to be like `ComponentDidMount`. When the page loads quickly this gets immediately set to true, but when it's taking a couple seconds to load (ie: comparing large diffs) I'm using this `loaded` state variable & useEffect to render the loading spinner until the whole component has loaded. I could rename the state variable to `isLoading` instead if that's more clear?

File: ggircs-app/app/pages/xml-diff.tsx (hunk 6 above)
Comment (dleard) [reply ggircs-app/app/pages/xml-diff.tsx:75]: do we? The leftsideReport / rightSideReport variables are complex objects which contain the full report xml

Comment (matthieu-foucault) [reply ggircs-app/app/pages/xml-diff.tsx:75]: the ids and boolean could be moved though, especially the ids, as it would enable sharing the URL, or refreshing the page, and still end up with the same view

File: ggircs-app/app/components/XmlDiff/RenderDiff.tsx (hunk 5 above)
Comment (matthieu-foucault) [reply ggircs-app/app/components/XmlDiff/RenderDiff.tsx:24]: I see. So this is simply to first render a loading spinner while the full, synchronous render is computing? This is not ideal, and we'd probably want to use a webworker for the diff calculation to keep the UI responsive, but that will do for now. Can you add a comment here to explain what this is for?

File: ggircs-app/app/pages/xml-diff.tsx (hunk 6 above)
Comment (dleard) [reply ggircs-app/app/pages/xml-diff.tsx]: variables to router work in issue #331

File: ggircs-app/app/next-env.d.ts (hunk 7 above)
Comment (dleard) [reply ggircs-app/app/next-env.d.ts:27]: Should these stay `static` or is pre-commit right? To keep them as `static` I can change them back and set pre-commit to ignore this file.

Comment (matthieu-foucault) [reply ggircs-app/app/next-env.d.ts:27]: try updating prettier and see if that changes anything? Otherwise if that's not causing any issue at compile time I think we can let pre-commit do its thing

Comment (dleard) [reply ggircs-app/app/next-env.d.ts:27]: updating prettier had no effect, leaving this as is.

**Review by matthieu-foucault — APPROVED** [review]
